librz/core/canalysis.c

```diff
@@ -6688,6 +6688,91 @@ RZ_IPI char *rz_core_analysis_function_signature(RzCore *core, RzOutputMode mode
 	return signature;
 }
 
+static RzAnalysisBlock *get_block_at(RzCore *core, ut64 addr) {
+	RzList *blocks = rz_analysis_get_blocks_in(core->analysis, addr);
+	if (!blocks) {
+		return NULL;
+	}
+	RzAnalysisBlock *block = NULL;
+	RzListIter *bit;
+	RzAnalysisBlock *block_cur;
+	rz_list_foreach (blocks, bit, block_cur) {
+		if (rz_analysis_block_op_starts_at(block_cur, addr)) {
+			block = block_cur;
+			break;
+		}
+	}
+	if (block) {
+		rz_analysis_block_ref(block);
+	}
+	rz_list_free(blocks);
+	return block;
+}
+
+RZ_API void rz_core_analysis_propagate_noreturn_relocs(RzCore *core, ut64 addr) {
+	// find known noreturn functions to propagate
+	RzListIter *iter;
+	RzAnalysisFunction *fcn;
+	RzList *noretl = rz_types_function_noreturn(core->analysis->sdb_types);
+	rz_list_foreach (core->analysis->fcns, iter, fcn) {
+		RzListIter *iter2;
+		RzList *refs = rz_analysis_function_get_refs(fcn);
+		RzAnalysisRef *ref;
+		rz_list_foreach (refs, iter2, ref) {
+			if (ref->type == RZ_ANALYSIS_REF_TYPE_CALL || ref->type == RZ_ANALYSIS_REF_TYPE_CODE) {
+				// At first we check if there are any relocations that override the call address
+				// Note, that the relocation overrides only the part of the instruction
+				ut64 addr = ref->at;
+				RzBinReloc *rel = rz_core_getreloc(core, addr, 8);
+				if (rel) {
+					RzListIter *iter3;
+					char *noret;
+					if (rel->import) {
+						rz_list_foreach (noretl, iter3, noret) {
+							if (!strcmp(rel->import->name, noret)) {
+								//eprintf("IMP: Adding noreturn for \"%s\" at 0x%08" PFMT64x "\n", rel->import->name, rel->vaddr);
+								fcn->is_noreturn = true;
+								ut8 buf[32] = {0};
+								RzAnalysisOp op = {0};
+								if (core->analysis->iob.read_at(core->analysis->iob.io, addr, buf, 32)) {
+									if (rz_analysis_op(core->analysis, &op, addr, buf, core->blocksize, 0)) {
+										//eprintf("IMP: Call size %d, chopping at 0x%08"PFMT64x"\n", op.size,
+										//		addr + op.size);
+										rz_analysis_noreturn_add(core->analysis, NULL, rel->vaddr);
+										// Chop the function
+										rz_analysis_function_chop_at(fcn, addr + op.size);
+									}
+								}
+							}
+						}
+					} else if (rel->symbol) {
+						rz_list_foreach (noretl, iter3, noret) {
+							if (!strcmp(rel->symbol->name, noret)) {
+								//eprintf("SYM: Adding noreturn for \"%s\" at 0x%08" PFMT64x "\n", rel->symbol->name, rel->symbol->vaddr);
+								fcn->is_noreturn = true;
```

Comment:
a lot of code duplication here

Reply:
Yes, I will refactor if the concept sounds OK for you.